ext/drreg/drreg.c

```diff
@@ -498,10 +498,20 @@ drreg_insert_restore_all(void *drcontext, instrlist_t *bb, instr_t *inst,
         }
     }
     /* Before each app read, or at end of bb, restore spilled registers to app values.
-     * XXX i#5121: multi-phase use causes correctness issues here!  Our restore
-     * in the insertion phase here could actually restore the tool value from the
-     * app2app phase.  A better solution is needed for multi-phase, or perhaps
-     * disallowing the use of the same register.
+     * XXX i#5121: For regs that are in spilled state at this instr also during the
+     * app2app phase, the following does not restore the app value, it restores the
+     * app2app meta value instead. This is expected and correct, because the drreg use
+     * in insertion phase adds spills/restores around the drreg spills/restores from
+     * previous phases like they were app instrs. So, the app2app meta value is no
+     * different than the app value for the insertion phase.
+     *
+     * For cases where we really want the app value at some point in the insertion phase,
+     * e.g. for use with annotations, or for user-added calls to drreg_get_app_value and
```

Comment:
But the app reading the register is even more important than annotations or drreg_get_app_value, and the user needs to restore for that too: so I also found this confusing.  The last sentence of this paragraph does say that but I don't understand why it's not considered a case "where we really want the app value".

Reply:
Oh I meant "where the instrumentation wants the app value"

Reply:
Resolved in a later comment thread.